### ISSUE TITLE
`SoftBody3D` issues from 0.13.0-beta1

### ISSUE THREAD

**Issue by Musicgun47** [godot-jolt/godot-jolt#909]
This is a list of the issues I have found so far in my testing of the new `SoftBody3D` implementation in the 0.13.0-beta1 release. Feel free to split them out into their own issues if necessary.

- ### Selecting vertices for pinning in the viewport doesn't work.
    - When clicking on vertices in the viewport, nothing happens. I am aware that this feature can be a little buggy with Godot Physics however, the vertices do get added to the pinned points even if they don't get highlighted. Currently clicking points doesn't add them to the pinned points array, even when deselecting and reselecting the soft body to refresh its rendering. This means currently the only way to add pinned points is manually via the inspector.
- ### Pinned Points array is buggy when modifying via the inspector window.
    - When attempting to modify the Pinned Points array in the inspector, the array will sometimes alter its size, deleting the last entry. I have found this happens consistently when rearranging the order of vertices, however there may be other times it happens also that I haven't been able to reliably reproduce.
- ### Simulation has a delay before starting.
    - When testing the simulation, I have found that the simulation doesn't seem to properly initialize until a collision is detected. I have tested this with a simple plane with one corner pinned hanging just above a static body floor. Gravity makes the plane swing down and collide with the floor however, it then just bounces like a piece of card before the simulation takes over and it folds as expected. This seems to happen regardless of the stiffness value.
- ### Soft Bodies with a skinned mesh don't render.
    - When creating a soft body from a skinned mesh (tested with a simple plane with a flag waving animation), the mesh doesn't render at all (points are still present and being moved by the animation as tested using a simple print command). I'm aware that this feature isn't supported by Godot Physics (and produces some 'interesting' results when used) so this may be a limitation of the interface, but I thought it should be mentioned anyway.

Not sure if this is the right way to report this as there doesn't seem to be any defined issue template, but this is what I've found so far. Also, I can upload a minimal test project/images if necessary.

**Comment by mihe** [comment]
> Selecting vertices for pinning in the viewport doesn't work.

> Pinned Points array is buggy when modifying via the inspector window.

I am painfully aware of these two, but they're not issues with this extension and aren't something I have control over. They should be reported to the Godot repository, if they're not already.

> Simulation has a delay before starting.

I'll look into this one. I have seen something to this effect before.

> Soft Bodies with a skinned mesh don't render.

I haven't tested this myself, but it doesn't surprise me that it wouldn't work. It's unlikely to be something I can fix from the extension side of things, but I'll take a look.

**Comment by Musicgun47** [comment]
> I am painfully aware of these two, but they're not something I have control over from this extension. They should be reported to the Godot repository, if they're not already.

Can confirm that these do indeed occur with Godot Physics as well and as far as I can see has not been reported yet, so I will definitely see about reporting that. This is also something I want to look into as part of fixing up the SoftBody UI as the current system is woefully inadequate for anything beyond simple uses. There is a proposal relating to fixing SoftBodies which mentions the interface, so I might head over there and have a chat.

As far as skinned meshes go, I sort of figured it would probably be outside of your control, but thought I'd mention it anyway. From what I can see, the mesh is getting lost somewhere along the way to the rendering server, so definitely an engine issue too.

**Comment by mihe** [comment]
> ### Simulation has a delay before starting.
> 
> - When testing the simulation, I have found that the simulation doesn't seem to properly initialize until a collision is detected. I have tested this with a simple plane with one corner pinned hanging just above a static body floor. Gravity makes the plane swing down and collide with the floor however, it then just bounces like a piece of card before the simulation takes over and it folds as expected. This seems to happen regardless of the stiffness value.

Apologies for taking so long to get back to this one, but would you say this reflects what you described above about the simulation having a delay?:

https://github.com/user-attachments/assets/a9a48202-8212-4258-8695-ae3f6841a4a9

Here's the minimal project I used, in case you wish to test it or tune it to better match what you experienced: [soft-body-cloth.zip](https://github.com/user-attachments/files/16575949/soft-body-cloth.zip)

**Comment by Musicgun47** [comment]
That's more or less it, although I have noticed that it's more prominent when it collides with the edge of the cloth (i.e. collision normal is roughly perpendicular to cloth normal). With the 4.3 stable release, I have noticed that it's not as prominent.

Also, I've found another issue where when the soft body goes to sleep, pinned points stop registering position updates. Basically, if my character stays still long enough for the cape to go to sleep, next time the character moves, the cape stays behind (see image below). I haven't added any of the collision shapes for the cape to the character yet, so that probably doesn't help, but may or may not be something to look into.

![image](https://github.com/user-attachments/assets/55cec912-5fc9-4664-8f40-ee4556f39099)

**Comment by mihe** [comment]
> That's more or less it

I'm not sure there's much that can be done about this type of oscillating behavior, at least not with the parameters exposed by Godot. Messing with the simulation precision doesn't seem to do much, and neither does damping (to my surprise) but maybe @jrouwe has some other ideas.

> although I have noticed that it's more prominent when it collides with the edge of the cloth

Would you be able to show a video of this?

> when the soft body goes to sleep, pinned points stop registering position updates

That is indeed a bug, and a major bummer. I wish I'd caught that before I published 0.13.0-stable.

**Comment by jrouwe** [comment]
> Messing with the simulation precision doesn't seem to do much, and neither does damping (to my surprise) but maybe @jrouwe has some other ideas.

I tried the `soft-body-cloth.zip` (cause that's what I think we're talking about right?). The reason it is oscillating is that there are quite a few vertices in that soft body. The iterative solver doesn't seem to get enough iterations to create a stable simulation. For me it works reasonably well if I up the 'Simulation Precision' to 10 (more iterations) and increase 'Damping Coefficient' to say 0.2 (0 is very low). What also helps is to reduce the number of subdivisions, so if I edit the PlaneMesh and set the subdivisions to 25 it oscillates a lot less.

Editing the code and setting `mCompliance` and `mShearCompliance` to 1.0e-6 (instead of 1.0e-5 which corresponds to 'Linear Stiffness' = 1) also helps, but this is borderline stable.

**Comment by mihe** [comment]
> cause that's what I think we're talking about right?

Yes!

> What also helps is to reduce the number of subdivisions, so if I edit the PlaneMesh and set the subdivisions to 25 it oscillates a lot less.

Right, yeah, this seemed to make the biggest difference for me.

Anyway, since that seemed to be mostly just about dialing in the settings right, and with this issue becoming a mixed bag of various bugs and not-really-bugs, I'll go ahead and refocus this one around the one bug (relevant to this extension) that was found, meaning the waking up being broken, and with that being fixed I'll go ahead and close this.
